scripts/tests/test_mnist_experiment.py

```
@@ -0,0 +1,90 @@
+                (staging_directory / "run.json").write_text("complete\n", encoding="utf-8")
+                self.assertFalse(output_directory.exists())
+
+            self.assertEqual((output_directory / "run.json").read_text(encoding="utf-8"), "complete\n")
+
+    def test_failed_run_does_not_publish_partial_artifacts(self) -> None:
+        """A failed generation leaves neither a final nor an incomplete run."""
+        with TemporaryDirectory() as temporary_directory:
+            root = Path(temporary_directory)
+            output_directory = root / "run"
+            message = "training failed"
+            with self.assertRaisesRegex(RuntimeError, message), _staged_run_directory(output_directory) as staging_directory:
```

Ruff (0.16.0)

[warning] Use `pytest.raises` instead of unittest-style `assertRaisesRegex`

Replace `assertRaisesRegex` with `pytest.raises`

(PT027)